openedx_blockstore_relay/data_transfer/unit.py

```diff
@@ -0,0 +1,116 @@
+        """
+        Store the block_key (UsageKey) and bundle_url
+        """
+        super(TransferUnit, self).__init__()
+        self.block_key = block_key
+        self.bundle_files_url = urljoin(bundle_url, '/'.join(['files', '']))
+
+    def sync_block_to_bundle(self):
+        """
+        Replace the Bundle content with the Unit components.
+
+        Intermediate Unit content is stored in memory.
+        """
+        LOG.debug('Transfer Unit %s to Bundle <%s>', self.block_key, self.bundle_files_url)
+        filesystem = MemoryFS()
+
+        # Can raise modulestore.exceptions.ItemNotFoundError
+        store = modulestore()
+        block = store.get_item(self.block_key)
+
+        xml_node = block.definition_to_xml(filesystem)
+        self.consolidate_olx(xml_node, filesystem)
+        self.update_bundle(filesystem)
+
+    def consolidate_olx(self, xml_node, filesystem):
+        """
+        Consolidate the generated unit content under the parent xml_node, and store a single OLX file if possible.
+
+        If the generated unit content doesn't parse as XML, then store as multiple files.
+        """
+        for item in filesystem.walk():
+            for unit_file in item.files:
+                file_path = path.join(str(item.path), unit_file.name)
+                file_contents = filesystem.gettext(file_path, encoding='utf-8')
+                try:
+                    # If we can parse the unit content, then append it to the node
+                    element = ElementTree.fromstring(file_contents)
+                    xml_node.append(element)
+
+                    # remove the html file we just internalised
+                    filesystem.remove(file_path)
+
+                    # and remove the "filename" attribute if it matches this file name.
+                    if xml_node.get('filename') == unit_file.name:
+                        xml_node.set('filename', None)
+
+                except ElementTree.ParseError as error:
+                    LOG.debug("Error parsing %s: code=%s, position=%s", file_path, error.code, error.position)
+
+        # TODO: scrape content for static files, and write them to the filesystem too
+
+        # TODO: generate a bundle.json manifest file
+
+        # Write the OLX file
+        olx_path = path.join(self.block_key.block_type, '.'.join([self.block_key.block_id, 'olx']))
+        with filesystem.open(olx_path, 'wb') as fp:
+            olx = ElementTree.tostring(xml_node, encoding="utf-8")
+            LOG.debug("OLX: %s", olx)
+            fp.write(olx)
+
+    def update_bundle(self, filesystem):
+        """
+        Walks the contents of the given filesystem and updates the given bundle with its content.
+
+        TODO consider requests.toolbelt to stream large file content: https://toolbelt.readthedocs.io/en/latest/
+        """
+        # FIXME requests says it is possible to post more than one file, but the bundle isn't ending up with the right
```

Comment:
The `POST` below fails with a `ValueError: too many values to unpack` when I run this code using a DemoX HTML block, e.g.:
```bash
./manage.py cms transfer_to_blockstore --settings=devstack --verbosity=3 --block "block-v1:edX+DemoX+Demo_Course+type@html+block@030e35c4756a4ddc8d40b95fbbfff4d4" --bundle "9a3bb1e0-463d-4972-9a06-a433b3dfd872"
```

debug log shows:
```
DEBUG|OLX: <html filename="030e35c4756a4ddc8d40b95fbbfff4d4" />
DEBUG|POST files=[(u'data', (u'030e35c4756a4ddc8d40b95fbbfff4d4.html', <fs.iotools.RawWrapper object at 0x7f7fe1b3c1d0>)), (u'data', (u'030e35c4756a4ddc8d40b95fbbfff4d4.olx', <fs.iotools.RawWrapper object at 0x7f7fe1b3c310>))], data=[{u'path': u'/html/030e35c4756a4ddc8d40b95fbbfff4d4.html', u'data': u'030e35c4756a4ddc8d40b95fbbfff4d4.html', u'public': True}, {u'path': u'/html/030e35c4756a4ddc8d40b95fbbfff4d4.olx', u'data': u'030e35c4756a4ddc8d40b95fbbfff4d4.olx', u'public': True}] to Bundle <http://localhost:8888/api/v1/bundles/9a3bb1e0-463d-4972-9a06-a433b3dfd872/files/>
```